File: src/types/analysis.ts
export type AnalysisInputSource = 'Demo case' | 'Uploaded file' | 'Pasted message'
export type AnalysisSeverity = 'Low' | 'Medium' | 'High' | 'Critical'
export type AnalysisStep = 'Uploading message' | 'Parsing email' | 'Analyzing evidence' | 'Building assessment' | 'Parsing message' | 'Inspecting headers' | 'Checking authentication' | 'Extracting URLs' | 'Inspecting attachments' | 'Evaluating threat indicators'

export interface EmailMetadata {
  from: string
  to: string
  subject: string
  date: string
  messageId: string
  attachmentCount: number
  urlCount: number
  replyTo?: string
  returnPath?: string
}

export interface DemoEmail {
  id: string
  title: string
  description: string
  threatLevel: AnalysisSeverity
  sender: string
  domain: string
  evidence: string
  metadata: EmailMetadata
  rawContent: string
}

export interface AuthenticationResult {
  spf: 'Pass' | 'Fail' | 'Neutral' | 'Unknown'
  dkim: 'Pass' | 'Fail' | 'Neutral' | 'Unknown'
  dmarc: 'Pass' | 'Fail' | 'Neutral' | 'Unknown'
}

export interface AnalysisUrl { value: string; verdict: 'Malicious' | 'Suspicious' | 'Clean' | 'Unknown'; reason?: string }
export interface AnalysisAttachment { name: string; type: string; verdict: 'Malicious' | 'Suspicious' | 'Clean' | 'Unknown'; size?: string; sha256?: string }
export interface AnalysisFinding { id?: string; category?: string; title: string; detail: string; evidence?: string; severity: AnalysisSeverity; riskContribution?: number }
export interface AnalysisIndicator { type: 'Domain' | 'IP address' | 'Hash' | 'URL' | 'Email'; value: string; source?: string; status?: 'Observed' | 'Review' | 'Clean' | 'Unknown' }
export interface AnalysisTimelineEvent { label: string; timestamp: string; detail: string }
export interface Geolocation { country: string; city: string; ip: string; region?: string; organization?: string; asn?: string; domain?: string }

export interface AnalysisResult {
  incidentId: string
  riskScore: number
  severity: AnalysisSeverity
  threatType: string
  metadata: EmailMetadata
  authentication: AuthenticationResult
  urls: AnalysisUrl[]
  attachments: AnalysisAttachment[]
  findings: AnalysisFinding[]
  indicators: AnalysisIndicator[]
  timeline: AnalysisTimelineEvent[]
  geolocation: Geolocation | null
  recommendations: string[]
  analyzedAt: string
  rawContent: string
}

File: backend/app/schemas/analysis.py
from pydantic import BaseModel
from typing import Literal

from app.schemas.email import EmailAttachment, EmailMetadata, EmailUrl


class HealthResponse(BaseModel):
    status: str
    service: str


class UploadAnalysisResponse(BaseModel):
    status: str
    filename: str
    content_type: str
    size_bytes: int


AnalysisSeverity = Literal["LOW", "MEDIUM", "HIGH", "CRITICAL"]
FindingCategory = Literal["authentication", "header", "url", "attachment", "content"]
AuthenticationState = Literal["PASS", "FAIL", "NEUTRAL", "UNKNOWN"]
UrlStatus = Literal["BENIGN", "SUSPICIOUS", "UNKNOWN"]
IndicatorType = Literal["EMAIL", "DOMAIN", "URL", "IP", "HASH"]
IndicatorStatus = Literal["OBSERVED", "SUSPICIOUS", "UNKNOWN"]


class Finding(BaseModel):
    id: str
    category: FindingCategory
    title: str
    severity: AnalysisSeverity
    description: str
    evidence: str
    risk_contribution: int
    source: str


class AuthenticationAnalysis(BaseModel):
    spf: AuthenticationState
    dkim: AuthenticationState
    dmarc: AuthenticationState
    dkim_signature_observed: bool = False


class AnalyzedUrl(BaseModel):
    url: str
    domain: str
    scheme: Literal["http", "https"]
    indicators: list[str]
    status: UrlStatus
    risk_contribution: int


class AnalyzedAttachment(BaseModel):
    filename: str
    content_type: str
    size_bytes: int
    sha256: str
    content_disposition: str | None = None
    indicators: list[str]
    status: Literal["OBSERVED", "SUSPICIOUS"]
    risk_contribution: int


class AnalysisIndicator(BaseModel):
    type: IndicatorType
    value: str
    source: str
    status: IndicatorStatus


class AnalysisTimelineEvent(BaseModel):
    label: str
    timestamp: str
    detail: str


class AnalysisResult(BaseModel):
    incident_id: str
    risk_score: int
    severity: AnalysisSeverity
    threat_type: str
    metadata: EmailMetadata
    authentication: AuthenticationAnalysis
    urls: list[AnalyzedUrl]
    attachments: list[AnalyzedAttachment]
    findings: list[Finding]
    indicators: list[AnalysisIndicator]
    timeline: list[AnalysisTimelineEvent]
    geolocation: None = None
    recommendations: list[str]
    analyzed_at: str
    forensic_footprint: dict[str, object] | None = None